### ISSUE TITLE
Asynchronous pulling of images

### ISSUE THREAD

**Issue by ayushjn20** [leopardslab/Dunner#32]
A single dunner file can contain multiple images whose containers will have many commands to run. Pulling images from a remote server (like Docker hub) one by one is quite slow. Concurrent pulling of images can be implemented using `goroutines`.

**Comment by agentmilindu** [comment]
This idea is really cool! But I have a little doubt? Will concurrent pulling saves the time, As the bandwidth is still constant? 

@rehrumesh @rhperera @tharindupr @charithccmc WDYT?

**Comment by rhperera** [comment]
This is nice work. I think we might need some options that one could specify to have control over the method of pulling images.

- Shall we have a switch, an option that we can tell Dunner if it should do the pulling parallel or sequential. If someone wants/have a single core, then they might want to turn off parallel pulling. docker-compose have the ability to define these options.

- Next one is, what is the maximum amount of parallel pulls that we should allow. Docker pulls have multiple download layers for each pull which will have multiple concurrent downloads. With bandwidth as a constraint, having too many concurrent downloads might have negative affect. docker-compose have a limit of 5 for max pulls per time and AFAIK docker-compose have --parallel mode enabled by default and it adjusts the number of concurrent pulls for the optimal amount.

Again nice work and IMO having the two options above would be nice

**Comment by ayushjn20** [comment]
Hey @agentmilindu @rhperera 
I actually did some analysis by modifying the code in main.go to calculate to the time elapsed. 

```
➜  Dunner git:(develop) ✗ go build . && ./Dunner do build
latest: Pulling from library/node
ab1fc7e4bf91: Pull complete 
35fba333ff52: Pull complete 
f0cb1fa13079: Pull complete 
3d1dd648b5ad: Pull complete 
49f7a0920ce1: Pull complete 
1d199f738c5f: Pull complete 
c9c43580a66c: Pull complete 
52af953a7504: Pull complete 
Digest: sha256:3a93398aa0ef09e5d61931f9d02b1b595fdf81204b550f870ac26999364b5329
Status: Downloaded newer image for node:latest
INFO[2019-01-25 20:51:35] Running task 'build' on 'node' Docker with command 'node --version' 
v11.7.0
latest: Pulling from library/node
Digest: sha256:3a93398aa0ef09e5d61931f9d02b1b595fdf81204b550f870ac26999364b5329
Status: Image is up to date for node:latest
INFO[2019-01-25 20:51:47] Running task 'build' on 'node' Docker with command 'npm --version' 
6.5.0
10.15.0: Pulling from library/node
ab1fc7e4bf91: Already exists 
35fba333ff52: Already exists 
f0cb1fa13079: Already exists 
3d1dd648b5ad: Already exists 
49f7a0920ce1: Already exists 
1d199f738c5f: Already exists 
669e5ed443bf: Pull complete 
d8dd16c377d0: Pull complete 
Digest: sha256:c9a51b7f2f3774f39058bb436ed62d66f415f32b9386e5c35c3fb4854722ed84
Status: Downloaded newer image for node:10.15.0
INFO[2019-01-25 20:52:04] Running task 'build' on 'node:10.15.0' Docker with command 'node --version' 
v10.15.0
10.15.0: Pulling from library/node
Digest: sha256:c9a51b7f2f3774f39058bb436ed62d66f415f32b9386e5c35c3fb4854722ed84
Status: Image is up to date for node:10.15.0
INFO[2019-01-25 20:52:11] Running task 'build' on 'node:10.15.0' Docker with command 'npm --version' 
6.4.1
latest: Pulling from library/alpine
cd784148e348: Already exists 
Digest: sha256:46e71df1e5191ab8b8034c5189e325258ec44ea739bba1e5645cff83c9048ff1
Status: Downloaded newer image for alpine:latest
INFO[2019-01-25 20:52:21] Running task 'build' on 'alpine' Docker with command 'apk update' 
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
v3.8.2-29-gbac7399976 [http://dl-cdn.alpinelinux.org/alpine/v3.8/main]
v3.8.2-28-g785c236c3e [http://dl-cdn.alpinelinux.org/alpine/v3.8/community]
OK: 9546 distinct packages available
Time taken: 1m59.804704706s%  
```

And this is the benchmark performance of my current work(making containerisation asynchronous too):
```
➜  Dunner git:(async-containerisation) ✗ go build . && ./Dunner do build
INFO[2019-01-25 20:44:53] Pulling image 'alpine'                       
INFO[2019-01-25 20:44:53] Pulling image 'node'                         
INFO[2019-01-25 20:44:53] Pulling image 'node:10.15.0'                 
INFO[2019-01-25 20:45:59] Pull complete
                               
INFO[2019-01-25 20:46:17] Running task 'build' on 'node' Docker with command 'npm --version' 
INFO[2019-01-25 20:46:18] Running task 'build' on 'node:10.15.0' Docker with command 'node --version' 
6.5.0
v10.15.0
INFO[2019-01-25 20:46:18] Running task 'build' on 'node' Docker with command 'node --version' 
INFO[2019-01-25 20:46:18] Running task 'build' on 'alpine' Docker with command 'apk update' 
INFO[2019-01-25 20:46:18] Running task 'build' on 'node:10.15.0' Docker with command 'npm --version' 
v11.7.0
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
v3.8.2-29-gbac7399976 [http://dl-cdn.alpinelinux.org/alpine/v3.8/main]
v3.8.2-28-g785c236c3e [http://dl-cdn.alpinelinux.org/alpine/v3.8/community]
OK: 9546 distinct packages available
6.4.1
Time taken 1m25.957357912s%
```
I have made sure to remove containers and images before re-running.

**Comment by ayushjn20** [comment]
@rhperera 

Implementing a switch requires incorporating flags using cobra's _Persistant Flags_ and global configurattion across the project using viper. Both of these are next in my pipeline. It would be really great if there is some progress over #33, so that I can make a new PR for asynchronous containerization. Then I will start working on the above-mentioned features.

**Comment by ayushjn20** [comment]
@agentmilindu @rhperera 
Have a look at pull #34. I hope the proposed implementation in the new PR is up to the mark.

**Comment by ayushjn20** [comment]
Closed with reference to #34
